### PR TITLE
Check queue status across all available servers

### PR DESCRIPTION
**Pull request by IdoBar**
Check queue status for user-submitted jobs across all queues in all available servers. 
This should fix issue #1840 
Please verify the escaping of the single quotes in line 85.

### CODE REVIEW

**Review by IdoBar — COMMENTED** [review]
This had been tested on Awoonga HPC running PBS Pro v18.1.2

**Comment by abhi18av** [comment]
@IdoBar thanks for this PR!

To make the `Travis CI` build happy regarding the tests, you'd need to update this test here https://github.com/nextflow-io/nextflow/blob/402339f6b054774375ef97abb86d95f87bfa18e6/modules/nextflow/src/test/groovy/nextflow/executor/PbsProExecutorTest.groovy#L150

**Comment by IdoBar** [comment]
@abhi18av, I've updated the test as well.

File: modules/nextflow/src/main/groovy/nextflow/executor/PbsProExecutor.groovy [hunk 1]
```diff
@@ -82,8 +82,8 @@ class PbsProExecutor extends PbsExecutor {
 
     @Override
     protected List<String> queueStatusCommand(Object queue) {
-        String cmd = 'qstat -f'
-        if( queue ) cmd += ' ' + queue
+        String cmd = 'qstat -u $USER -f $( qstat -B | egrep -v \'(^Server|^---)\' | awk -v ORS=" " \'{print "@"$1}\' )'
+       // if( queue ) cmd += ' ' + queue
```
Comment (IdoBar) [inline modules/nextflow/src/main/groovy/nextflow/executor/PbsProExecutor.groovy:86]:
```suggestion
        String cmd = 'qstat -u $USER -f ' 
        if( queue ) {
            cmd += queue
        } else {
            cmd += '$( qstat -B | egrep -v \'(^Server|^---)\' | awk -v ORS=" " \'{print "@"$1}\' )'
        }
```

**Comment by IdoBar** [comment]
@pditommaso, @abhi18av, thanks for your feedback, it took me a few iterations, but I managed to pass the tests.

**Comment by pditommaso** [comment]
What exactly `$( qstat -B | egrep -v \'(^Server|^---)\' | awk -v ORS=\' \' \'{print \"@\"\$1}\' )` does ?

**Comment by IdoBar** [comment]
> 
> 
> What exactly `$( qstat -B | egrep -v \'(^Server|^---)\' | awk -v ORS=\' \' \'{print \"@\"\$1}\' )` does ?

It parses all the available servers for the scheduler and makes a list of arguments for `qstat`, so on my HPC, when jobs are automatically redirected to multiple queues at different servers, it will run `qstat -u $USER -f @server1 @server2 @server3` and will capture all the jobs for that user.
I left the original command as is if the user supplies a specific `queue` parameter.

**Comment by pditommaso** [comment]
Is this portable on other PBS cluster? you mentioned you are using a modified version.

**Comment by IdoBar** [comment]
> 
> 
> Is this portable on other PBS clusters? you mentioned you are using a modified version.

Yes, as I've mentioned above it's working as intended on other PBS clusters, whether they support automatic job redirection to other servers or not (it uses "regular" `qstat` commands and I've tested it on another cluster that I have access to).
With this patch I won't need to use my modified version of `qstat` (which is the whole purpose of it).

File: modules/nextflow/src/main/groovy/nextflow/executor/PbsProExecutor.groovy [hunk 2]
```diff
@@ -82,8 +82,12 @@ class PbsProExecutor extends PbsExecutor {
 
     @Override
     protected List<String> queueStatusCommand(Object queue) {
-        String cmd = 'qstat -f'
-        if( queue ) cmd += ' ' + queue
+        String cmd = 'qstat -u $USER -f '
```
Comment (IdoBar) [inline modules/nextflow/src/main/groovy/nextflow/executor/PbsProExecutor.groovy:85]:
```suggestion
        String cmd = 'qstat -f ' 
```

File: modules/nextflow/src/test/groovy/nextflow/executor/PbsProExecutorTest.groovy [hunk 3]
```diff
@@ -150,10 +150,11 @@ class PbsProExecutorTest extends Specification {
...
         given:
         def executor = [:] as PbsProExecutor
-
+		def user = System.getenv('USER')
+		
         expect:
-        executor.queueStatusCommand(null) == ['bash','-c', "set -o pipefail; qstat -f | { egrep '(Job Id:|job_state =)' || true; }"]
-        executor.queueStatusCommand('xxx') == ['bash','-c', "set -o pipefail; qstat -f xxx | { egrep '(Job Id:|job_state =)' || true; }"]
+        executor.queueStatusCommand('xxx') == ['bash','-c', "set -o pipefail; qstat -u \$USER -f xxx | { egrep '(Job Id:|job_state =)' || true; }"]
+		executor.queueStatusCommand(null) == ['bash','-c', "set -o pipefail; qstat -u \$USER -f \$( qstat -B | egrep -v '(^Server|^---)' | awk -v ORS=' ' '{print \"@\"\$1}' ) | { egrep '(Job Id:|job_state =)' || true; }"]
```
Comment (IdoBar) [inline modules/nextflow/src/test/groovy/nextflow/executor/PbsProExecutorTest.groovy:157]:
```suggestion
        executor.queueStatusCommand('xxx') == ['bash','-c', "set -o pipefail; qstat -f xxx | { egrep '(Job Id:|job_state =)' || true; }"]
		executor.queueStatusCommand(null) == ['bash','-c', "set -o pipefail; qstat -f \$( qstat -B | egrep -v '(^Server|^---)' | awk -v ORS=' ' '{print \"@\"\$1}' ) | { egrep '(Job Id:|job_state =)' || true; }"]
```

**Comment by pditommaso** [comment]
I noticed the `-u` option has gone. Is this expected to be the final version?

**Comment by IdoBar** [comment]
@pditommaso Yes, this version is tested and working as intended on several HPCs running PBSPro and I'll try to test it on a few others.

**Comment by pditommaso** [comment]
@arnaudceol @cjfields If I'm not wrong you heavily really on the PSB pro executor for Nextflow. Could you validate the proposed change in this PR to make that works in your cluster?   

Essentially it consists of using the following command line to fetch the queue(s) status 

```
set -o pipefail; $( qstat -B | egrep -v '(^Server|^---)' | awk -v ORS=' ' '{print "@"$1}' ) | { egrep '(Job Id:|job_state =)' || true; }
```

instead of 

```
set -o pipefail; qstat -f <queue name> | { egrep '(Job Id:|job_state =)' || true; }
```

Thanks a lot

**Comment by cjfields** [comment]
@pditommaso would love to help but our local HPC switched primarily to SLURM a bit ago.  I can check with our NCSA colleagues, I think they may be running PBS Pro on some systems.

**Comment by pditommaso** [comment]
I see, at your convenience! thanks a lot for your swift reply!

**Comment by IdoBar** [comment]
> 
> 
> @arnaudceol @cjfields If I'm not wrong you heavily really on the PSB pro executor for Nextflow. Could you validate the proposed change in this PR to make that works in your cluster?
> 
> Essentially it consists of using the following command line to fetch the queue(s) status
> 
> ```
> set -o pipefail; $( qstat -B | egrep -v '(^Server|^---)' | awk -v ORS=' ' '{print "@"$1}' ) | { egrep '(Job Id:|job_state =)' || true; }
> ```
> 
> instead of
> 
> ```
> set -o pipefail; qstat -f <queue name> | { egrep '(Job Id:|job_state =)' || true; }
> ```
> 
> Thanks a lot

Just a correction, , the command still starts with `qstat -f`: 
```
set -o pipefail; qstat -f $( qstat -B | egrep -v '(^Server|^---)' | awk -v ORS=' ' '{print "@"$1}' ) | { egrep '(Job Id:|job_state =)' || true; }
```

**Comment by pditommaso** [comment]
My fault, I've misread it. Updated the comment.

**Comment by arnaudceol** [comment]
@pditommaso  I confirm that the command 

```
set -o pipefail; qstat -f $( qstat -B | egrep -v '(^Server|^---)' | awk -v ORS=' ' '{print "@"$1}' ) | { egrep '(Job Id:|job_state =)' || true; }
```

returns the list jobs on all queues (we have only one server, pbs_version = 19.1.2).

**Comment by IdoBar** [comment]
I can also confirm that it works as intended on the HPC at QUT, running PBS version 2020.1.0.20200720152739

**Comment by pditommaso** [comment]
Nice. Thanks a lot both! My only concern is the amount of output this command can produce checking all queues. In a busy cluster, it can be thousands of tasks. What it's preventing limiting `qstat` to the submitting user?

**Comment by IdoBar** [comment]
> 
> 
> Nice. Thanks a lot both! My only concern is the amount of output this command can produce checking all queues. In a busy cluster, it can be thousands of tasks. What it's preventing limiting `qstat` to the submitting user?

`qstat -f` doesn't work with `-u`, these 2 flags are mutually exclusive (except for some reason on our specific HPC cluster).
If a specific queue is provided, however, then only the jobs of that queue will be listed (as it was previously).
I think that since each job ID is collected at submission, it would be best to check only the status of these jobs with `qstat -f $JOB_ID1 $JOB_ID2 ...`.
Alternatively, the running status of the jobs can be parsed from the short version of `qstat -u $USER`, without the `-f` flag (which is must easier to parse) - see an example of the output:
```
> qstat -u $USER
gc-prd-hpcadm:
                                                            Req'd  Req'd   Elap
Job ID          Username Queue    Jobname    SessID NDS TSK Memory Time  S Time
--------------- -------- -------- ---------- ------ --- --- ------ ----- - -----
79184.gc-prd-hp s5000942 workq    KimJob      49510   1  16   36gb 300:0 R 276:4
79974.gc-prd-hp s5000942 workq    KimJob      67429   1   4   24gb 300:0 R 245:3
80145.gc-prd-hp s5000942 workq    KimJob     177233   1   6   24gb 300:0 R 236:0
81011.gc-prd-hp s5000942 workq    KimJob     118637   1   4   24gb 200:0 R 196:3
81407.gc-prd-hp s5000942 workq    KimJob      38131   1   4   24gb 200:0 R 172:5
81408.gc-prd-hp s5000942 workq    KimJob      36075   1   4   24gb 200:0 R 172:5
81680.gc-prd-hp s5000942 workq    KimJob      49214   1  16   24gb 500:0 R 71:33
81683.gc-prd-hp s5000942 workq    KimJob      37961   1   4   24gb 300:0 R 71:25
81695.gc-prd-hp s5000942 workq    KimJob      44326   1   4   24gb 300:0 R 71:21
81699.gc-prd-hp s5000942 workq    KimJob     166262   1   4   24gb 300:0 R 71:02
81700.gc-prd-hp s5000942 workq    KimJob     149223   1   4   24gb 300:0 R 70:52
81703.gc-prd-hp s5000942 workq    KimJob     165458   1   4   24gb 300:0 R 70:38
81877.gc-prd-hp s5000942 workq    KimJob      72323   1   4   24gb 300:0 R 27:59
```
That can be easily parsed to extract the Job ID and Job State with the following
```
> qstat -u $USER | grep "$USER" | awk '{print $1,$10}'
79184.gc-prd-hp R
79974.gc-prd-hp R
80145.gc-prd-hp R
81011.gc-prd-hp R
81407.gc-prd-hp R
81408.gc-prd-hp R
81680.gc-prd-hp R
81683.gc-prd-hp R
81695.gc-prd-hp R
81699.gc-prd-hp R
81700.gc-prd-hp R
81703.gc-prd-hp R
81877.gc-prd-hp R
````

I think that will be much faster and more focused than parsing the entire server(s) or entire queue, but I just need to verify that the parsing command will work on my cluster when jobs are spread across multiple servers (can't see a reason why it wouldn't).

**Comment by pditommaso** [comment]
The -x flag is very useful and I agree that the `qstat -xu` would be a better solution, the problem is that we need to validate across different installations.

**Comment by arnaudceol** [comment]
> The -x flag is very useful and I agree that the `qstat -xu` would be a better solution, the problem is that we need to validate across different installations.

qstat -xu is one I use on our cluster. Be careful, it may provide really long lists. It shows the jobs on all queues (cannot confirm on a multi-server installation).

To get more info about the jobs (if it may help), -ant can be used:

```
# qstat -u MyID

hpcfe01: 
                                                            Req'd  Req'd   Elap
Job ID          Username Queue    Jobname    SessID NDS TSK Memory Time  S Time
--------------- -------- -------- ---------- ------ --- --- ------ ----- - -----
436184.hpcfe01  MyID  workq    pathseq.sh 121545   1   2   40gb   --  R 67:40




# qstat -antu MyID

hpcfe01: 
                                                            Req'd  Req'd   Elap
Job ID          Username Queue    Jobname    SessID NDS TSK Memory Time  S Time
--------------- -------- -------- ---------- ------ --- --- ------ ----- - -----
436184.hpcfe01  MyID  workq    pathseq.sh 121545   1   2   40gb   --  R 67:40
   cn01/3*2
```

**Comment by pditommaso** [comment]
I guess the list gets long when specifying the `-x` because it contains the completed jobs? is so, is there any flag to specify how old complete jobs can be e.g. last hour?

**Comment by IdoBar** [comment]
Is there an actual need to list completed jobs?
At the moment it doesn't list these jobs either with `qstat -f`, so I guess it can stay that way.

**Comment by pditommaso** [comment]
> At the moment it doesn't list these jobs either with qstat -f, so I guess it can stay that way.

That's correct, however, having the ability to check the status of completed jobs would allow the better handing of killed jobs. When the status is not detected from the `qstat` command NF fallback checking on the `.exitcode` file. But this file cannot be created if the job is hard killed by the cluster.

In any case, I'm fine with this patch, since it's only used when the `queue` is not specified.

**Comment by pditommaso** [comment]
Merged manually. 👉  c159890. Thanks for your contribution.

**Pull request closed without merging**